mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td

```diff
@@ -577,13 +579,28 @@ def LLVM_DISubprogramAttr : LLVM_Attr<"DISubprogram", "di_subprogram",
       "ArrayRef<DINodeAttr>":$retainedNodes
     ), [{
       MLIRContext *ctx = file.getContext();
-      return $_get(ctx, id, compileUnit, scope, StringAttr::get(ctx, name),
+      return $_get(ctx, id, /*recId=*/nullptr, compileUnit, scope,
+                   StringAttr::get(ctx, name),
                    StringAttr::get(ctx, linkageName), file, line,
                    scopeLine, subprogramFlags, type, retainedNodes);
     }]>
   ];
-
   let assemblyFormat = "`<` struct(params) `>`";
+  let extraClassDeclaration = [{
+    /// Requirements of DIRecursiveTypeAttrInterface.
+    /// @{
+
+    /// Get whether this attr describes a recursive self reference.
+    bool isRecSelf() { return !getScope(); }
```

Comment:
My only slight concern is if it's possible to see a DISubprogram without a scope naturally (it looks like it's allowed to be null [here](https://github.com/llvm/llvm-project/blob/9ce4af5cadc24060f3c3674e01902d374afea983/llvm/lib/IR/Verifier.cpp#L1443) but I'm not sure if it's illegal/impractical elsewhere...). If it's not supposed to happen, can we enforce during import that all llvm's DISubprograms have a scope?

The concern is if we categorize a node as RecSelf when it isn't, we won't be able to tell the outer real node from the inner recursive reference (which triggers an assertion during export).

Reply:
Alternatively we could maybe add a tag parameter to the enum similar to the one we have on DICompositeType? It is a bit redundant though since a subprogram is a subprogram. So maybe a separate boolean makes sense.

It is indeed true that using the scope feels a bit hacky.

Reply:
Yeah looking back it was pure luck we had a tag in DICompositeTypeAttr for this 😂 I guess adding an enum tag or a bool flag probably works the same. A tag certainly increases the textual IR a lot... Or, worst case we just make `recId` itself a dedicated attr with DistinctAttr + bool (may be handy if we need it for more nodes later).